src/power.jl

```diff
@@ -58,37 +58,98 @@ end
 
 ^(a::Taylor1{TaylorN{T}}, r::Rational) where {T<:NumberNotSeries} = a^(r.num/r.den)
 
-
-
-# power_by_squaring; slightly modified from base/intfuncs.jl
-# Licensed under MIT "Expat"
-for T in (:Taylor1, :HomogeneousPolynomial, :TaylorN)
-    @eval function power_by_squaring(x::$T, p::Integer)
-        if p == 1
-            return copy(x)
-        elseif p == 0
-            return one(x)
-        elseif p == 2
-            return square(x)
-        end
+# in-place form of power_by_squaring
+# this method assumes `y`, `x` and `aux` are of same order
+for T in (:Taylor1, :TaylorN)
+    @eval function power_by_squaring!(y::$T{T}, x::$T{T}, aux::$T{T},
+            p::Integer) where {T<:NumberNotSeries}
         t = trailing_zeros(p) + 1
         p >>= t
+        # aux = x
+        for k in eachindex(aux)
+            identity!(aux, x, k)
+        end
         while (t -= 1) > 0
-            x = square(x)
+            # aux = square(aux)
+            for k in reverse(eachindex(aux))
+                sqr!(aux, k)
+            end
+        end
+        # y = aux
+        for k in eachindex(y)
+            identity!(y, aux, k)
         end
-        y = x
         while p > 0
             t = trailing_zeros(p) + 1
             p >>= t
             while (t -= 1) ≥ 0
+                # aux = square(aux)
+                for k in reverse(eachindex(aux))
+                    sqr!(aux, k)
+                end
+            end
+            # y = y * aux
+            mul!(y, aux)
+        end
+        return nothing
+    end
+end
+
+
+# power_by_squaring; slightly modified from base/intfuncs.jl
+# Licensed under MIT "Expat"
+for T in (:Taylor1, :HomogeneousPolynomial, :TaylorN)
+    @eval power_by_squaring(x::$T, p::Integer) = power_by_squaring(x, Val(p))
+    @eval power_by_squaring(x::$T, ::Val{0}) = one(x)
+    @eval power_by_squaring(x::$T, ::Val{1}) = copy(x)
+    @eval power_by_squaring(x::$T, ::Val{2}) = square(x)
+    @eval function power_by_squaring(x::$T, ::Val{P}) where P
+        p = P # copy static parameter `P` into local variable `p`
+        if $T != HomogeneousPolynomial
+            y = zero(x)
+            aux = zero(x)
+            power_by_squaring!(y, x, aux, p)
+        else
+            t = trailing_zeros(p) + 1
+            p >>= t
+            while (t -= 1) > 0
                 x = square(x)
             end
-            y *= x
+            y = x
+            while p > 0
+                t = trailing_zeros(p) + 1
+                p >>= t
+                while (t -= 1) ≥ 0
+                    x = square(x)
+                end
+                y *= x
+            end
         end
```

Comment:
I think this `if` block is now unnecessary, and can be replaced simply by:
```suggestion
        y = zero(x)
        aux = zero(x)
        power_by_squaring!(y, x, aux, p)
```

Reply:
Thank you for noticing this @lbenet! I agree there's better ways to deal with this `if` block, I'm just wondering if it makes sense to have a `power_by_squaring!` for `HomogeneousPolynomial`, since inside such methods, there's a lot of in-place squaring going around, which would require a different treatment, since squaring a `HomogeneousPolynomial` changes its order, so some extra auxiliaries might be needed, I think? Anyway, I'm pushing a commit where the `power_by_squaring!` method for `HomogeneousPolynomial` is declared separately, so that this `if` does not occur inside the method.

Reply:
I agree it is worth keeping *both* methods for `HomogeneousPolynomial`s. I think they can be useful in methods that exploit `evaluate`, which is worth checking, though perhaps in another PR.

Reply:
Okay, I added a comment with a TODO for another PR